### PR TITLE
[FLINK-12888] [table-planner-blink] Introduce planner rule to push filter into TableSource

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Introduce planner rule to push filter into TableSource*

## Brief change log

  - *remove FunctionCatalog in blink planner and use FunctionCatalog in flink-table-api-java*
  - *Added PushFilterIntoTableSourceScanRule that pushes filter into TableSource*
  - *Added RexNodeExtractor to extract conjunctive predicates to Expressions*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added PushFilterIntoTableSourceScanRuleTest that validates logical plan after PushFilterIntoTableSourceScanRule is applied*
  - *Added TableSourceTest that validates the physical plan for the filterable table source*
  - *Added TableSourceITCase that validates the execution result for the filterable table source*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java [hunk 1]
```diff
@@ -168,6 +170,12 @@ private RexNode visitScalarFunc(UnresolvedCallExpression unresolvedCall) {
 			return relBuilder.call(FlinkSqlOperatorTable.MULTIPLY, child);
 		} else if (BuiltInFunctionDefinitions.MOD.equals(def)) {
 			return relBuilder.call(FlinkSqlOperatorTable.MOD, child);
+		} else if (def instanceof ScalarFunctionDefinition) {
+			// TODO add getName for ScalarFunctionDefinition ???
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java]: It looks wrong code, should be:
```
ScalarFunction scalaFunc = ((ScalarFunctionDefinition) func).getScalarFunction();
			List<RexNode> child = convertCallChildren(call);
			SqlFunction sqlFunction = UserDefinedFunctionUtils.createScalarSqlFunction(
					scalaFunc.functionIdentifier(),
					scalaFunc.toString(),
					scalaFunc,
					typeFactory);
			return relBuilder.call(sqlFunction, child);
```

Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java]: thanks for the correction. 
`List<RexNode> child = convertCallChildren(call);` is defined at line 122

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java]: Now we just use `functionIdentifier` to be name.
The detail reason is in https://github.com/apache/flink/pull/3330

Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java]: we can use `functionIdentifier` now, and fix it in another issue

**Comment by godfreyhe** [comment]
@KurtYoung, @JingsongLi,  i have updated this pr based on comments. do you have any other suggestions? thanks

ps: i add another minor commit into this pr: remove FunctionCatalog in blink planner and use FunctionCatalog in flink-table-api-java

**Pull request closed without merging**